script/generate-metadata-for-builder.js

```diff
@@ -0,0 +1,209 @@
+      msg => {
+        if (!msg.match(/No README data$/)) {
+          console.warn(
+            `Invalid package metadata. ${packageMetadata.name}: ${msg}`
+          );
+        }
+      },
+      true
+    );
+    if (
+      packageMetadata.repository &&
+      packageMetadata.repository.url &&
+      packageMetadata.repository.type === 'git'
+    ) {
+      packageMetadata.repository.url = packageMetadata.repository.url.replace(
+        /^git\+/,
+        ''
+      );
+    }
+
+    delete packageMetadata['_from'];
+    delete packageMetadata['_id'];
+    delete packageMetadata['dist'];
+    delete packageMetadata['readme'];
+    delete packageMetadata['readmeFilename'];
+
+    const packageModuleCache = packageMetadata._atomModuleCache || {};
+    if (
+      packageModuleCache.extensions &&
+      packageModuleCache.extensions['.json']
+    ) {
+      const index = packageModuleCache.extensions['.json'].indexOf(
+        'package.json'
+      );
+      if (index !== -1) {
+        packageModuleCache.extensions['.json'].splice(index, 1);
+      }
+    }
+
+    const packageNewMetadata = {
+      metadata: packageMetadata,
+      keymaps: {},
+      menus: {},
+      grammarPaths: [],
+      settings: {}
+    };
+
+    packageNewMetadata.rootDirPath = packagePath;
+
+    if (packageMetadata.main) {
+      const mainPath = require.resolve(
+        path.resolve(packagePath, packageMetadata.main)
+      );
+      packageNewMetadata.main = path.relative(
+        'static',
+        mainPath
+      );
+      // Convert backward slashes to forward slashes in order to allow package
+      // main modules to be required from the snapshot. This is because we use
+      // forward slashes to cache the sources in the snapshot, so we need to use
+      // them here as well.
+      packageNewMetadata.main = packageNewMetadata.main.replace(/\\/g, '/');
+    }
+
+    const packageKeymapsPath = path.join(packagePath, 'keymaps');
+    if (fs.existsSync(packageKeymapsPath)) {
+      for (let packageKeymapName of fs.readdirSync(packageKeymapsPath)) {
+        const packageKeymapPath = path.join(
+          packageKeymapsPath,
+          packageKeymapName
+        );
+        if (
+          packageKeymapPath.endsWith('.cson') ||
+          packageKeymapPath.endsWith('.json')
+        ) {
+          packageNewMetadata.keymaps[packageKeymapPath] = CSON.readFileSync(
+            packageKeymapPath
+          );
+        }
+      }
+    }
+
+    const packageMenusPath = path.join(packagePath, 'menus');
+    if (fs.existsSync(packageMenusPath)) {
+      for (let packageMenuName of fs.readdirSync(packageMenusPath)) {
+        const packageMenuPath = path.join(packageMenusPath, packageMenuName);
+        if (
+          packageMenuPath.endsWith('.cson') ||
+          packageMenuPath.endsWith('.json')
+        ) {
+          packageNewMetadata.menus[packageMenuPath] = CSON.readFileSync(
+            packageMenuPath
+          );
+        }
+      }
+    }
```

Comment:
Each of these blocks of code can become a function with a descriptive name and good documentation.